File: frontend/src/services/tms.ts
/**
 * TMS API Service - 任务管理系统前端服务
 */
import api from './api';

// ============== Types ==============

export interface TMSTask {
  id: string;
  task_code: string;
  title: string;
  description?: string;
  task_type: string;
  source: string;
  priority: 'low' | 'medium' | 'high' | 'urgent';
  points: number;
  status: string;
  distribution_strategy?: string;
  assigned_to?: string;
  assigned_by?: string;
  candidate_pool: Array<{ user_id: string; score?: number }>;
  required_skills: string[];
  required_roles: string[];
  deadline?: string;
  approval_flow_id?: string;
  agent_context: Record<string, any>;
  metadata: Record<string, any>;
  created_by?: string;
  created_at: string;
  updated_at: string;
  // 推荐任务附加字段
  match_rate?: number;
  ai_recommendation?: string;
}

export interface DashboardStats {
  pending_distribution: number;
  distributed: number;
  claimed: number;
  in_progress: number;
  pending_approval: number;
  completed: number;
  rejected: number;
  total: number;
  weekly_points: number;
  sla_rate: number;
}

export interface PendingApproval {
  flow_id: string;
  flow_code: string;
  task_id: string;
  task_code: string;
  task_title: string;
  task_type: string;
  priority: string;
  current_step: number;
  step_name: string;
  initiated_by?: string;
  created_at?: string;
}

export interface AgentCommandResponse {
  success: boolean;
  command: string;
  data: Record<string, any>;
  message: string;
  requires_confirmation: boolean;
  confirmation_id?: string;
  action_id?: string;
}

export interface DistributionStats {
  status_distribution: Record<string, number>;
  strategy_usage: Record<string, number>;
  total_distributions: number;
}

// ============== Task API ==============

export const tmsApi = {
  // Tasks
  createTask: (data: {
    title: string;
    task_type?: string;
    description?: string;
    priority?: string;
    points?: number;
    required_skills?: string[];
    deadline?: string;
  }) => api.post('/api/v1/tms/tasks', data),

  listTasks: (params?: {
    status?: string;
    task_type?: string;
    priority?: string;
    assigned_to?: string;
    page?: number;
    page_size?: number;
  }) => api.get('/api/v1/tms/tasks', { params }),

  getTask: (taskId: string) => api.get(`/api/v1/tms/tasks/${taskId}`),

  updateTask: (taskId: string, data: Partial<TMSTask>) =>
    api.put(`/api/v1/tms/tasks/${taskId}`, data),

  // Distribution
  distributeTask: (taskId: string, data: {
    strategy?: string;
    mode?: string;
    target_user_id?: string;
  }) => api.post(`/api/v1/tms/tasks/${taskId}/distribute`, data),

  claimTask: (taskId: string, userId: string) =>
    api.post(`/api/v1/tms/tasks/${taskId}/claim`, { user_id: userId }),

  getDistributionStats: () => api.get('/api/v1/tms/distribution/stats'),

  // Approvals
  initiateApproval: (data: {
    task_id: string;
    flow_type?: string;
    steps?: Array<Record<string, any>>;
  }) => api.post('/api/v1/tms/approvals', data),

  approveTask: (flowId: string, data: { approver_id: string; comment?: string }) =>
    api.post(`/api/v1/tms/approvals/${flowId}/approve`, data),

  rejectTask: (flowId: string, data: { approver_id: string; comment?: string }) =>
    api.post(`/api/v1/tms/approvals/${flowId}/reject`, data),

  delegateApproval: (flowId: string, data: { from_user_id: string; to_user_id: string }) =>
    api.post(`/api/v1/tms/approvals/${flowId}/delegate`, data),

  escalateApproval: (flowId: string, data: { reason: string }) =>
    api.post(`/api/v1/tms/approvals/${flowId}/escalate`, data),

  getPendingApprovals: (approverId: string) =>
    api.get('/api/v1/tms/approvals/pending', { params: { approver_id: approverId } }),

  getApprovalFlow: (flowId: string) => api.get(`/api/v1/tms/approvals/${flowId}`),

  // Agent API
  agentCommand: (data: {
    agent_id: string;
    command: string;
    params: Record<string, any>;
    idempotency_key?: string;
  }) => api.post('/api/v1/tms/agent/command', data),

  confirmAgentAction: (data: {
    action_id: string;
    confirmed_by: string;
    approved?: boolean;
  }) => api.post('/api/v1/tms/agent/confirm', data),

  registerAgent: (data: {
    agent_id: string;
    permission_level?: number;
    whitelisted?: boolean;
  }) => api.post('/api/v1/tms/agent/register', data),

  registerWebhook: (data: {
    agent_id: string;
    event_types: string[];
    webhook_url: string;
    secret?: string;
  }) => api.post('/api/v1/tms/agent/webhook', data),

  getAgentContext: (taskId: string) => api.get(`/api/v1/tms/agent/context/${taskId}`),

  // Dashboard
  getDashboardStats: () => api.get('/api/v1/tms/dashboard/stats'),

  getRecommendedTasks: (userId: string, limit?: number) =>
    api.get('/api/v1/tms/dashboard/recommended', { params: { user_id: userId, limit } }),
};

export default tmsApi;

File: api/schemas/tms.py
"""
TMS API Schemas - Pydantic 请求/响应模型
"""

from datetime import datetime
from typing import Optional, List, Dict, Any
from pydantic import BaseModel, Field


# ============== Task Schemas ==============

class TaskCreate(BaseModel):
    """创建任务请求"""
    title: str = Field(..., max_length=200, description="任务标题")
    description: Optional[str] = Field(None, description="任务描述")
    task_type: str = Field(default="custom", description="任务类型: ecn_release, ecr_approval, inspection, custom")
    priority: str = Field(default="medium", description="优先级: low, medium, high, urgent")
    points: int = Field(default=0, ge=0, description="积分")
    required_skills: List[str] = Field(default_factory=list, description="所需技能")
    required_roles: List[str] = Field(default_factory=list, description="所需角色")
    deadline: Optional[datetime] = Field(None, description="截止日期")
    distribution_strategy: Optional[str] = Field(None, description="分发策略")
    related_work_order_id: Optional[str] = Field(None, description="关联工单ID")
    metadata: Dict[str, Any] = Field(default_factory=dict, description="扩展元数据")


class TaskUpdate(BaseModel):
    """更新任务请求"""
    title: Optional[str] = None
    description: Optional[str] = None
    priority: Optional[str] = None
    points: Optional[int] = None
    required_skills: Optional[List[str]] = None
    deadline: Optional[datetime] = None
    metadata: Optional[Dict[str, Any]] = None


class TaskResponse(BaseModel):
    """任务响应"""
    id: str
    task_code: str
    title: str
    description: Optional[str]
    task_type: str
    source: str
    priority: str
    points: int
    status: str
    distribution_strategy: Optional[str]
    assigned_to: Optional[str]
    assigned_by: Optional[str]
    candidate_pool: List[Dict[str, Any]]
    required_skills: List[str]
    required_roles: List[str]
    deadline: Optional[datetime]
    approval_flow_id: Optional[str]
    agent_context: Dict[str, Any]
    metadata: Dict[str, Any]
    created_by: Optional[str]
    created_at: datetime
    updated_at: datetime

    class Config:
        from_attributes = True


class TaskListResponse(BaseModel):
    """任务列表响应"""
    items: List[TaskResponse]
    total: int
    page: int
    page_size: int
    total_pages: int


class TaskDistributeRequest(BaseModel):
    """任务分发请求"""
    strategy: str = Field(default="skill_match", description="分发策略")
    mode: str = Field(default="direct", description="分发模式: direct, pool, agent")
    target_user_id: Optional[str] = Field(None, description="手动指定分配对象")


class TaskClaimRequest(BaseModel):
    """任务认领请求"""
    user_id: str = Field(..., description="认领人ID")


# ============== Approval Schemas ==============

class ApprovalFlowCreate(BaseModel):
    """创建审批流请求"""
    task_id: str = Field(..., description="关联任务ID")
    flow_type: str = Field(default="sequential", description="流类型: sequential, parallel, conditional")
    steps: Optional[List[Dict[str, Any]]] = Field(None, description="审批步骤定义")


class ApprovalActionRequest(BaseModel):
    """审批操作请求"""
    approver_id: str = Field(..., description="审批人ID")
    comment: Optional[str] = Field(None, description="审批意见")


class ApprovalDelegateRequest(BaseModel):
    """委托审批请求"""
    from_user_id: str = Field(..., description="委托人ID")
    to_user_id: str = Field(..., description="被委托人ID")


class ApprovalEscalateRequest(BaseModel):
    """升级审批请求"""
    reason: str = Field(..., description="升级原因")


class ApprovalFlowResponse(BaseModel):
    """审批流响应"""
    flow_id: str
    flow_code: str
    task_id: str
    flow_type: str
    status: str
    current_step: int
    total_steps: int
    steps: List[Dict[str, Any]]
    records: List[Dict[str, Any]]


class PendingApprovalItem(BaseModel):
    """待审批项"""
    flow_id: str
    flow_code: str
    task_id: str
    task_code: str
    task_title: str
    task_type: str
    priority: str
    current_step: int
    step_name: str
    initiated_by: Optional[str]
    created_at: Optional[str]


# ============== Agent Schemas ==============

class AgentCommandRequest(BaseModel):
    """Agent 命令请求"""
    agent_id: str = Field(..., description="Agent 标识")
    command: str = Field(..., description="命令名称")
    params: Dict[str, Any] = Field(default_factory=dict, description="命令参数")
    idempotency_key: Optional[str] = Field(None, description="幂等键")


class AgentCommandResponse(BaseModel):
    """Agent 命令响应"""
    success: bool
    command: str
    data: Dict[str, Any]
    message: str
    requires_confirmation: bool = False
    confirmation_id: Optional[str] = None
    action_id: Optional[str] = None


class AgentConfirmRequest(BaseModel):
    """Agent 操作确认请求"""
    action_id: str = Field(..., description="待确认操作ID")
    confirmed_by: str = Field(..., description="确认人ID")
    approved: bool = Field(default=True, description="是否批准")


class AgentRegisterRequest(BaseModel):
    """Agent 注册请求"""
    agent_id: str = Field(..., description="Agent 标识")
    permission_level: int = Field(default=1, ge=1, le=3, description="权限等级 1-3")
    whitelisted: bool = Field(default=False, description="是否加入白名单")


class WebhookRegisterRequest(BaseModel):
    """Webhook 注册请求"""
    agent_id: str = Field(..., description="Agent 标识")
    event_types: List[str] = Field(..., description="订阅事件类型列表")
    webhook_url: str = Field(..., description="Webhook URL")
    secret: Optional[str] = Field(None, description="签名密钥")


# ============== Distribution Schemas ==============

class DistributionLogResponse(BaseModel):
    """分发日志响应"""
    id: str
    task_id: str
    strategy: str
    candidate_scores: Dict[str, Any]
    selected_user_id: Optional[str]
    reason: Optional[str]
    triggered_by: str
    created_at: datetime


class DistributionStatsResponse(BaseModel):
    """分发统计响应"""
    status_distribution: Dict[str, int]
    strategy_usage: Dict[str, int]
    total_distributions: int


# ============== Dashboard Schemas ==============

class TMSDashboardStats(BaseModel):
    """TMS 仪表盘统计"""
    pending_distribution: int = 0
    distributed: int = 0
    claimed: int = 0
    in_progress: int = 0
    pending_approval: int = 0
    completed: int = 0
    rejected: int = 0
    total: int = 0
    weekly_points: int = 0
    sla_rate: float = 0.0


class TaskCardItem(BaseModel):
    """任务卡片项（前端展示用）"""
    task_id: str
    task_code: str
    title: str
    task_type: str
    priority: str
    points: int
    status: str
    assigned_to_name: Optional[str]
    deadline: Optional[str]
    required_skills: List[str]
    ai_recommendation: Optional[str] = None
    risk_alert: Optional[str] = None
